File: web/tests/gc-flow.test.js
/**
 * Headless tests for web/js/lib/gc-flow.js (WP 4a.4).
 *
 * Mirrors the Android sibling's `GcFlowTest.kt` (WP 4b.6) case for case —
 * "GC execute NEVER sent without an explicit user confirm after a dry
 * run... pin the state machine cannot reach execute from idle".
 *
 * Run: node --test "web/tests/*.test.js"   (see web/tests/README.md)
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import { GC_STATE, GC_EVENT, idleGcState, reduceGcFlow } from "../js/lib/gc-flow.js";

function job(id, status, log_excerpt = null) {
  return { id, appid: 440, type: "gc", status, created_at: "2026-08-01T00:00:00Z", log_excerpt };
}

// ---- the flagship pin -------------------------------------------------

test("THE PIN -- confirm_execute against idle stays idle, execute can never be reached from idle", () => {
  const result = reduceGcFlow(idleGcState(), { type: GC_EVENT.CONFIRM_EXECUTE });
  assert.deepEqual(result, idleGcState());
});

test("confirm_execute is rejected from every state except confirm_execute itself", () => {
  const nonConfirmStates = [
    idleGcState(),
    { kind: GC_STATE.REQUESTING_DRY_RUN },
    { kind: GC_STATE.POLLING_DRY_RUN, jobId: 1 },
    { kind: GC_STATE.DRY_RUN_PLAN, jobId: 1, job: job(1, "done"), summary: null },
    { kind: GC_STATE.REQUESTING_EXECUTE },
    { kind: GC_STATE.POLLING_EXECUTE, jobId: 1 },
    { kind: GC_STATE.EXECUTE_DONE, jobId: 1, job: job(1, "done"), summary: null },
    { kind: GC_STATE.CANCELLED, jobId: 1 },
    { kind: GC_STATE.ERROR, message: "boom", executeAttempted: false },
  ];
  for (const state of nonConfirmStates) {
    assert.deepEqual(reduceGcFlow(state, { type: GC_EVENT.CONFIRM_EXECUTE }), state);
  }
});

test("request_execute is rejected from every state except dry_run_plan (full parametrised pin)", () => {
  // The reachable regression this specifically guards: from execute_done (a
  // GC that already ran to completion), a second "Execute" tap must NOT
  // re-confirm the OLD plan and queue a second execute without a fresh dry
  // run -- it must be a no-op, exactly like every other non-dry_run_plan state.
  const nonDryRunPlanStates = [
    idleGcState(),
    { kind: GC_STATE.REQUESTING_DRY_RUN },
    { kind: GC_STATE.POLLING_DRY_RUN, jobId: 1 },
    { kind: GC_STATE.CONFIRM_EXECUTE, jobId: 1, job: job(1, "done"), summary: null },
    { kind: GC_STATE.REQUESTING_EXECUTE },
    { kind: GC_STATE.POLLING_EXECUTE, jobId: 1 },
    { kind: GC_STATE.EXECUTE_DONE, jobId: 1, job: job(1, "done"), summary: null },
    { kind: GC_STATE.CANCELLED, jobId: 1 },
    { kind: GC_STATE.ERROR, message: "boom", executeAttempted: false },
  ];
  for (const state of nonDryRunPlanStates) {
    assert.deepEqual(reduceGcFlow(state, { type: GC_EVENT.REQUEST_EXECUTE }), state);
  }
});

// ---- the full valid path ------------------------------------------------

test("full dry-run then execute path", () => {
  let state = idleGcState();

  state = reduceGcFlow(state, { type: GC_EVENT.START_DRY_RUN });
  assert.deepEqual(state, { kind: GC_STATE.REQUESTING_DRY_RUN });

  state = reduceGcFlow(state, { type: GC_EVENT.DRY_RUN_QUEUED, jobId: 7 });
  assert.deepEqual(state, { kind: GC_STATE.POLLING_DRY_RUN, jobId: 7 });

  state = reduceGcFlow(state, { type: GC_EVENT.POLL_RESULT, job: job(7, "queued") });
  assert.deepEqual(state, { kind: GC_STATE.POLLING_DRY_RUN, jobId: 7 }); // still in flight

  state = reduceGcFlow(state, { type: GC_EVENT.POLL_RESULT, job: job(7, "running") });
  assert.deepEqual(state, { kind: GC_STATE.POLLING_DRY_RUN, jobId: 7 });

  const doneJob = job(7, "done", "[vault-api] GC totals (DRY RUN): would_delete=2 (700 bytes)");
  state = reduceGcFlow(state, { type: GC_EVENT.POLL_RESULT, job: doneJob });
  assert.equal(state.kind, GC_STATE.DRY_RUN_PLAN);
  assert.equal(state.jobId, 7);
  assert.equal(state.summary.wouldDeleteCount, 2);

  state = reduceGcFlow(state, { type: GC_EVENT.REQUEST_EXECUTE });
  assert.equal(state.kind, GC_STATE.CONFIRM_EXECUTE);

  // A dismiss returns to the plan without ever touching the network.
  const dismissed = reduceGcFlow(state, { type: GC_EVENT.DISMISS_CONFIRM });
  assert.equal(dismissed.kind, GC_STATE.DRY_RUN_PLAN);

  // The actual confirm.
  state = reduceGcFlow(state, { type: GC_EVENT.CONFIRM_EXECUTE });
  assert.deepEqual(state, { kind: GC_STATE.REQUESTING_EXECUTE });

  state = reduceGcFlow(state, { type: GC_EVENT.EXECUTE_QUEUED, jobId: 8 });
  assert.deepEqual(state, { kind: GC_STATE.POLLING_EXECUTE, jobId: 8 });

  const executedJob = job(
    8,
    "done",
    "[vault-api] GC totals (EXECUTED): chunks_removed=2 bytes_freed=700 total_bytes_freed=700",
  );
  state = reduceGcFlow(state, { type: GC_EVENT.POLL_RESULT, job: executedJob });
  assert.equal(state.kind, GC_STATE.EXECUTE_DONE);
  assert.equal(state.jobId, 8);
  assert.equal(state.summary.chunksRemoved, 2);
});

// ---- error paths ----------------------------------------------------------

test("a dry-run request that fails to even queue reports executeAttempted = false", () => {
  const state = reduceGcFlow({ kind: GC_STATE.REQUESTING_DRY_RUN }, { type: GC_EVENT.DRY_RUN_FAILED, message: "network error" });
  assert.deepEqual(state, { kind: GC_STATE.ERROR, message: "network error", executeAttempted: false });
});

test("an execute request that fails to even queue reports executeAttempted = true", () => {
  const state = reduceGcFlow({ kind: GC_STATE.REQUESTING_EXECUTE }, { type: GC_EVENT.EXECUTE_FAILED, message: "network error" });
  assert.deepEqual(state, { kind: GC_STATE.ERROR, message: "network error", executeAttempted: true });
});

test("a job that reaches error status while polling the dry run becomes Error(executeAttempted=false)", () => {
  const state = reduceGcFlow({ kind: GC_STATE.POLLING_DRY_RUN, jobId: 1 }, { type: GC_EVENT.POLL_RESULT, job: job(1, "error") });
  assert.deepEqual(state, { kind: GC_STATE.ERROR, message: "GC job 1 failed", executeAttempted: false });
});

test("a job that reaches error status while polling the execute becomes Error(executeAttempted=true)", () => {
  const state = reduceGcFlow({ kind: GC_STATE.POLLING_EXECUTE, jobId: 1 }, { type: GC_EVENT.POLL_RESULT, job: job(1, "error") });
  assert.deepEqual(state, { kind: GC_STATE.ERROR, message: "GC job 1 failed", executeAttempted: true });
});

// ---- cancellation mid-poll ------------------------------------------------

test("a job cancelled by another client mid dry-run poll becomes Cancelled, not Error", () => {
  const state = reduceGcFlow({ kind: GC_STATE.POLLING_DRY_RUN, jobId: 1 }, { type: GC_EVENT.POLL_RESULT, job: job(1, "cancelled") });
  assert.deepEqual(state, { kind: GC_STATE.CANCELLED, jobId: 1 });
});

test("a job cancelled by another client mid execute poll becomes Cancelled, not Error", () => {
  const state = reduceGcFlow({ kind: GC_STATE.POLLING_EXECUTE, jobId: 1 }, { type: GC_EVENT.POLL_RESULT, job: job(1, "cancelled") });
  assert.deepEqual(state, { kind: GC_STATE.CANCELLED, jobId: 1 });
});

test("after Cancelled, start_dry_run works again -- Cancelled is a terminal-but-restartable state", () => {
  const state = reduceGcFlow({ kind: GC_STATE.CANCELLED, jobId: 1 }, { type: GC_EVENT.START_DRY_RUN });
  assert.deepEqual(state, { kind: GC_STATE.REQUESTING_DRY_RUN });
});

// ---- stray/stale poll results never corrupt state --------------------------

test("a poll result for a DIFFERENT job id than the one being tracked is ignored -- dry run", () => {
  const state = { kind: GC_STATE.POLLING_DRY_RUN, jobId: 1 };
  const result = reduceGcFlow(state, { type: GC_EVENT.POLL_RESULT, job: job(999, "done") });
  assert.deepEqual(result, state);
});

test("a poll result for a DIFFERENT job id than the one being tracked is ignored -- execute", () => {
  const state = { kind: GC_STATE.POLLING_EXECUTE, jobId: 7 };
  const result = reduceGcFlow(state, { type: GC_EVENT.POLL_RESULT, job: job(999, "done") });
  assert.deepEqual(result, state);
});

test("a poll result landing outside any polling state is ignored", () => {
  assert.deepEqual(
    reduceGcFlow(idleGcState(), { type: GC_EVENT.POLL_RESULT, job: job(1, "done") }),
    idleGcState(),
  );
});

test("an unrecognized job status keeps polling rather than guessing", () => {
  assert.deepEqual(
    reduceGcFlow({ kind: GC_STATE.POLLING_DRY_RUN, jobId: 1 }, { type: GC_EVENT.POLL_RESULT, job: job(1, "paused") }),
    { kind: GC_STATE.POLLING_DRY_RUN, jobId: 1 },
  );
  assert.deepEqual(
    reduceGcFlow({ kind: GC_STATE.POLLING_EXECUTE, jobId: 1 }, { type: GC_EVENT.POLL_RESULT, job: job(1, "paused") }),
    { kind: GC_STATE.POLLING_EXECUTE, jobId: 1 },
  );
});

// ---- reset ------------------------------------------------------------------

test("reset always returns to idle from any state", () => {
  assert.deepEqual(reduceGcFlow({ kind: GC_STATE.POLLING_EXECUTE, jobId: 1 }, { type: GC_EVENT.RESET }), idleGcState());
  assert.deepEqual(
    reduceGcFlow({ kind: GC_STATE.ERROR, message: "x", executeAttempted: true }, { type: GC_EVENT.RESET }),
    idleGcState(),
  );
  assert.deepEqual(reduceGcFlow(idleGcState(), { type: GC_EVENT.RESET }), idleGcState());
});

// ---- start_dry_run gating -----------------------------------------------------

test("start_dry_run is rejected while a flow is already mid-run", () => {
  const midFlight = [
    { kind: GC_STATE.REQUESTING_DRY_RUN },
    { kind: GC_STATE.POLLING_DRY_RUN, jobId: 1 },
    { kind: GC_STATE.DRY_RUN_PLAN, jobId: 1, job: job(1, "done"), summary: null },
    { kind: GC_STATE.CONFIRM_EXECUTE, jobId: 1, job: job(1, "done"), summary: null },
    { kind: GC_STATE.REQUESTING_EXECUTE },
    { kind: GC_STATE.POLLING_EXECUTE, jobId: 1 },
  ];
  for (const state of midFlight) {
    assert.deepEqual(reduceGcFlow(state, { type: GC_EVENT.START_DRY_RUN }), state);
  }
});

test("start_dry_run is accepted from idle, execute_done and error too", () => {
  assert.deepEqual(reduceGcFlow(idleGcState(), { type: GC_EVENT.START_DRY_RUN }), { kind: GC_STATE.REQUESTING_DRY_RUN });
  assert.deepEqual(
    reduceGcFlow({ kind: GC_STATE.EXECUTE_DONE, jobId: 1, job: job(1, "done"), summary: null }, { type: GC_EVENT.START_DRY_RUN }),
    { kind: GC_STATE.REQUESTING_DRY_RUN },
  );
  assert.deepEqual(
    reduceGcFlow({ kind: GC_STATE.ERROR, message: "x", executeAttempted: false }, { type: GC_EVENT.START_DRY_RUN }),
    { kind: GC_STATE.REQUESTING_DRY_RUN },
  );
});

File: app/app/src/test/java/dev/steamvault/app/ui/detail/logic/GcFlowTest.kt
package dev.steamvault.app.ui.detail.logic

import dev.steamvault.app.net.model.JobDetail
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Pins the GC flow's state machine (WP 4b.6 brief: "GC flow state machine
 * (dry-run -> plan-shown -> executing -> done/error, incl. cancellation
 * mid-poll)... GC execute NEVER sent without an explicit user confirm after
 * a dry-run -- pin the state machine cannot reach execute from idle").
 */
class GcFlowTest {

    private fun job(id: Int, status: String, log: String? = null) = JobDetail(
        id = id, appid = 440, type = "gc", status = status, created_at = "2026-08-01T00:00:00Z", log_excerpt = log,
    )

    // ---- the flagship pin ---------------------------------------------------

    @Test
    fun `THE PIN -- ConfirmExecute against Idle stays Idle, execute can never be reached from idle`() {
        val result = reduceGcFlow(GcFlowState.Idle, GcFlowEvent.ConfirmExecute)
        assertEquals(GcFlowState.Idle, result)
    }

    @Test
    fun `ConfirmExecute is rejected from every state except ConfirmExecute itself`() {
        val nonConfirmStates = listOf(
            GcFlowState.Idle,
            GcFlowState.RequestingDryRun,
            GcFlowState.PollingDryRun(1),
            GcFlowState.DryRunPlan(1, job(1, "done"), null),
            GcFlowState.RequestingExecute,
            GcFlowState.PollingExecute(1),
            GcFlowState.ExecuteDone(1, job(1, "done"), null),
            GcFlowState.Cancelled(1),
            GcFlowState.Error("boom", executeAttempted = false),
        )
        for (state in nonConfirmStates) {
            assertEquals("ConfirmExecute must be a no-op from $state", state, reduceGcFlow(state, GcFlowEvent.ConfirmExecute))
        }
    }

    @Test
    fun `S2 -- RequestExecute is rejected from every state except DryRunPlan (full parametrised pin)`() {
        // Mirrors the ConfirmExecute-no-op test above's thoroughness
        // (review fix S2). The reachable regression this specifically
        // guards: from ExecuteDone (a GC that already ran to completion), a
        // second "Execute" tap must NOT re-confirm the OLD plan and queue a
        // second execute without a fresh dry run -- it must be a no-op,
        // exactly like every other non-DryRunPlan state.
        val nonDryRunPlanStates = listOf(
            GcFlowState.Idle,
            GcFlowState.RequestingDryRun,
            GcFlowState.PollingDryRun(1),
            GcFlowState.ConfirmExecute(1, job(1, "done"), null),
            GcFlowState.RequestingExecute,
            GcFlowState.PollingExecute(1),
            GcFlowState.ExecuteDone(1, job(1, "done"), null),
            GcFlowState.Cancelled(1),
            GcFlowState.Error("boom", executeAttempted = false),
        )
        for (state in nonDryRunPlanStates) {
            assertEquals("RequestExecute must be a no-op from $state", state, reduceGcFlow(state, GcFlowEvent.RequestExecute))
        }
    }

    // ---- the full valid path --------------------------------------------------

    @Test
    fun `full dry-run then execute path`() {
        var state: GcFlowState = GcFlowState.Idle

        state = reduceGcFlow(state, GcFlowEvent.StartDryRun)
        assertEquals(GcFlowState.RequestingDryRun, state)

        state = reduceGcFlow(state, GcFlowEvent.DryRunQueued(7))
        assertEquals(GcFlowState.PollingDryRun(7), state)

        state = reduceGcFlow(state, GcFlowEvent.PollResult(job(7, "queued")))
        assertEquals(GcFlowState.PollingDryRun(7), state) // still in flight

        state = reduceGcFlow(state, GcFlowEvent.PollResult(job(7, "running")))
        assertEquals(GcFlowState.PollingDryRun(7), state)

        val doneJob = job(7, "done", log = "[vault-api] GC totals (DRY RUN): would_delete=2 (700 bytes)")
        state = reduceGcFlow(state, GcFlowEvent.PollResult(doneJob))
        assertTrue(state is GcFlowState.DryRunPlan)
        state as GcFlowState.DryRunPlan
        assertEquals(7, state.jobId)
        assertEquals(2, state.summary?.wouldDeleteCount)

        state = reduceGcFlow(state, GcFlowEvent.RequestExecute)
        assertTrue(state is GcFlowState.ConfirmExecute)

        // A dismiss returns to the plan without ever touching the network.
        val dismissed = reduceGcFlow(state, GcFlowEvent.DismissConfirm)
        assertTrue(dismissed is GcFlowState.DryRunPlan)

        // The actual confirm.
        state = reduceGcFlow(state, GcFlowEvent.ConfirmExecute)
        assertEquals(GcFlowState.RequestingExecute, state)

        state = reduceGcFlow(state, GcFlowEvent.ExecuteQueued(8))
        assertEquals(GcFlowState.PollingExecute(8), state)

        val executedJob = job(8, "done", log = "[vault-api] GC totals (EXECUTED): chunks_removed=2 bytes_freed=700 total_bytes_freed=700")
        state = reduceGcFlow(state, GcFlowEvent.PollResult(executedJob))
        assertTrue(state is GcFlowState.ExecuteDone)
        state as GcFlowState.ExecuteDone
        assertEquals(8, state.jobId)
        assertEquals(2, state.summary?.chunksRemoved)
    }

    // ---- error paths ------------------------------------------------------------

    @Test
    fun `a dry-run request that fails to even queue reports executeAttempted = false`() {
        var state: GcFlowState = GcFlowState.RequestingDryRun
        state = reduceGcFlow(state, GcFlowEvent.DryRunFailed("network error"))
        assertEquals(GcFlowState.Error("network error", executeAttempted = false), state)
    }

    @Test
    fun `an execute request that fails to even queue reports executeAttempted = true`() {
        var state: GcFlowState = GcFlowState.RequestingExecute
        state = reduceGcFlow(state, GcFlowEvent.ExecuteFailed("network error"))
        assertEquals(GcFlowState.Error("network error", executeAttempted = true), state)
    }

    @Test
    fun `a job that reaches error status while polling the dry run becomes Error(executeAttempted=false)`() {
        val state = reduceGcFlow(GcFlowState.PollingDryRun(1), GcFlowEvent.PollResult(job(1, "error")))
        assertEquals(GcFlowState.Error("GC job 1 failed", executeAttempted = false), state)
    }

    @Test
    fun `a job that reaches error status while polling the execute becomes Error(executeAttempted=true)`() {
        val state = reduceGcFlow(GcFlowState.PollingExecute(1), GcFlowEvent.PollResult(job(1, "error")))
        assertEquals(GcFlowState.Error("GC job 1 failed", executeAttempted = true), state)
    }

    // ---- cancellation mid-poll ------------------------------------------------------

    @Test
    fun `a job cancelled by another client mid dry-run poll becomes Cancelled, not Error`() {
        val state = reduceGcFlow(GcFlowState.PollingDryRun(1), GcFlowEvent.PollResult(job(1, "cancelled")))
        assertEquals(GcFlowState.Cancelled(1), state)
    }

    @Test
    fun `a job cancelled by another client mid execute poll becomes Cancelled, not Error`() {
        val state = reduceGcFlow(GcFlowState.PollingExecute(1), GcFlowEvent.PollResult(job(1, "cancelled")))
        assertEquals(GcFlowState.Cancelled(1), state)
    }

    @Test
    fun `after Cancelled, StartDryRun works again -- Cancelled is a terminal-but-restartable state`() {
        val state = reduceGcFlow(GcFlowState.Cancelled(1), GcFlowEvent.StartDryRun)
        assertEquals(GcFlowState.RequestingDryRun, state)
    }

    // ---- stray/stale poll results never corrupt state --------------------------------

    @Test
    fun `a poll result for a DIFFERENT job id than the one being tracked is ignored -- DRY RUN`() {
        val state = GcFlowState.PollingDryRun(1)
        val result = reduceGcFlow(state, GcFlowEvent.PollResult(job(999, "done")))
        assertEquals(state, result)
    }

    @Test
    fun `S1 -- a poll result for a DIFFERENT job id than the one being tracked is ignored -- EXECUTE`() {
        // Mirrors the dry-run twin above (review fix S1: this binding was
        // unpinned for PollingExecute -- removing the `job.id != state.jobId`
        // check in reducePoll's PollingExecute branch survived the full
        // suite unnoticed). Without it, a stale poll for some OTHER job
        // (e.g. a superseded execute run, or a completely unrelated GC job)
        // landing while this flow is polling ITS OWN execute job would be
        // applied as if it were the real outcome.
        val state = GcFlowState.PollingExecute(7)
        val result = reduceGcFlow(state, GcFlowEvent.PollResult(job(999, "done")))
        assertEquals(state, result)
    }

    @Test
    fun `a poll result landing outside any Polling state is ignored`() {
        assertEquals(GcFlowState.Idle, reduceGcFlow(GcFlowState.Idle, GcFlowEvent.PollResult(job(1, "done"))))
    }

    @Test
    fun `an unrecognized job status keeps polling rather than guessing`() {
        assertEquals(GcFlowState.PollingDryRun(1), reduceGcFlow(GcFlowState.PollingDryRun(1), GcFlowEvent.PollResult(job(1, "paused"))))
        assertEquals(GcFlowState.PollingExecute(1), reduceGcFlow(GcFlowState.PollingExecute(1), GcFlowEvent.PollResult(job(1, "paused"))))
    }

    // ---- Reset ------------------------------------------------------------------------

    @Test
    fun `Reset always returns to Idle from any state`() {
        assertEquals(GcFlowState.Idle, reduceGcFlow(GcFlowState.PollingExecute(1), GcFlowEvent.Reset))
        assertEquals(GcFlowState.Idle, reduceGcFlow(GcFlowState.Error("x", true), GcFlowEvent.Reset))
        assertEquals(GcFlowState.Idle, reduceGcFlow(GcFlowState.Idle, GcFlowEvent.Reset))
    }

    // ---- StartDryRun gating -----------------------------------------------------------

    @Test
    fun `StartDryRun is rejected while a flow is already mid-run`() {
        val midFlight = listOf(
            GcFlowState.RequestingDryRun,
            GcFlowState.PollingDryRun(1),
            GcFlowState.DryRunPlan(1, job(1, "done"), null),
            GcFlowState.ConfirmExecute(1, job(1, "done"), null),
            GcFlowState.RequestingExecute,
            GcFlowState.PollingExecute(1),
        )
        for (state in midFlight) {
            assertEquals("StartDryRun must be a no-op from $state", state, reduceGcFlow(state, GcFlowEvent.StartDryRun))
        }
    }

    @Test
    fun `StartDryRun is accepted from Idle, ExecuteDone and Error too`() {
        assertEquals(GcFlowState.RequestingDryRun, reduceGcFlow(GcFlowState.Idle, GcFlowEvent.StartDryRun))
        assertEquals(
            GcFlowState.RequestingDryRun,
            reduceGcFlow(GcFlowState.ExecuteDone(1, job(1, "done"), null), GcFlowEvent.StartDryRun),
        )
        assertEquals(
            GcFlowState.RequestingDryRun,
            reduceGcFlow(GcFlowState.Error("x", false), GcFlowEvent.StartDryRun),
        )
    }
}
